### PR TITLE
[DTensor] Handle None gradients in from_local and to_local backward functions

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #173295
* __->__ #173865

When an autograd function has multiple outputs but only some are used, the gradients for unused
outputs are None. By default, autograd materializes these as zero tensors.

https://github.com/pytorch/pytorch/pull/173295 improved this for DTensor outputs by using zeros_like() to preserve the DTensor type.  However, https://github.com/pytorch/pytorch/pull/173295 doesn't work with from_local(). When autograd materializes a zero gradient for from_local, it creates a plain torch.Tensor, which causes a type mismatch error when mixed with DTensor in subsequent computations.

This PR updates both from_local and to_local to explicitly handle None gradients in their backward
functions, avoiding reliance on autograd's default materialization behavior.

### CODE REVIEW

File: torch/distributed/tensor/_api.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def forward(  # type: ignore[override]
     ):
         ctx.dtensor_spec = input._spec
         ctx.grad_placements = grad_placements
+        ctx.set_materialize_grads(False)
```
Comment (wconstab) [inline torch/distributed/tensor/_api.py:101]: this just tells autograd "We can handle none grad_out"?

File: test/distributed/tensor/test_dtensor.py [hunk 2]
```diff
@@ -698,6 +698,30 @@ def func(x, y):
...
+                return (x * 2), (y * 3)
+
+            @staticmethod
+            def backward(ctx, grad_out1, grad_out2):
+                assert grad_out2 is None
+                return grad_out1 * 2, grad_out2
+
+        x_local = torch.randn(4, 4, device=self.device_type, requires_grad=True)
+        y_local = torch.randn(4, 4, device=self.device_type, requires_grad=True)
+        x = DTensor.from_local(x_local, device_mesh, [Replicate()])
```
Comment (wconstab) [inline test/distributed/tensor/test_dtensor.py:845]: i wonder why autograd didn't give us zeros_like dtensor for from_local grad_out?  from_local returns a DTensor during forward out.  isn't that the thing that autograd would look at when doing zeros_like?

(for to_local, i would expect the opposite: it returns a regular tensor output, so grad should also be a regular tensor.

maybe i am misunderstanding how the autograd handling works.

File: torch/distributed/tensor/_api.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/tensor/_api.py:101]: yes

File: test/distributed/tensor/test_dtensor.py (hunk 2 above)
Comment (fegin) [reply test/distributed/tensor/test_dtensor.py:845]: hmm, you are right, then it should be the `to_local()` that is wrong. Not quite so sure. It seems that https://github.com/pytorch/pytorch/pull/173295 should at least work with `from_local()`. This only happens when the downstream is also an `autograd.Function` and set `set_materialize_grads(False)`. @soulitzer do you why https://github.com/pytorch/pytorch/pull/173295  doesn't fix `from_local()`?

Comment (fegin) [reply test/distributed/tensor/test_dtensor.py:845]: oh, I know why. The autograd only use DTensor.zeros_like() if num_outputs > 1. This logic exist when using zeros_like for NJT. I think it is safe for DTensor to remove the "num_outputs > 1" condition, but we can preserve the original logic to play safe since we fix the to_local and from_local in this PR. 

cc., @wconstab @soulitzer

**Review by soulitzer — APPROVED** [review]

**Review by wconstab — APPROVED** [review]
